src/ucp/core/ucp_worker.c

```diff
@@ -790,12 +793,37 @@ void ucp_worker_iface_unprogress_ep(ucp_worker_iface_t *wiface)
     UCS_ASYNC_UNBLOCK(&wiface->worker->async);
 }
 
-void ucp_worker_iface_event(int fd, void *arg)
+static UCS_F_ALWAYS_INLINE void
+ucp_worker_iface_event_common(ucp_worker_iface_t *wiface)
 {
-    ucp_worker_iface_t *wiface = arg;
     ucp_worker_h worker = wiface->worker;
+
+    /* Do more work on the main thread */
+    ucp_worker_iface_check_events(wiface, 0);
+
+    /* Signal user wakeup to report the first message on the interface */
+    ucp_worker_signal_internal(worker);
+}
+
+static ucs_status_t ucp_worker_iface_event_async_cb(void *arg,
+                                                    unsigned flags)
+{
+    ucp_worker_iface_t *wiface = arg;
+
+    ucs_assert(wiface->attr.cap.event_flags & UCT_IFACE_FLAG_EVENT_ASYNC_CB);
+    ucs_trace_func("async_cb for iface=%p", wiface->iface);
+
+    ucp_worker_iface_event_common(wiface);
+
+    return UCS_OK;
+}
+
+static void ucp_worker_iface_event(int fd, void *arg)
```

Comment:
maybe rename the UCP callbacks as follows:
FD: ucp_worker_iface_async_fd_event
CB: ucp_worker_iface_async_cb_event

Reply:
done